File: agent/packages/schema/src/poi-novice.ts
import { Type, type Static } from "@sinclair/typebox";

import { validate, type ValidationResult } from "./validate.js";

export const PoiNoviceKindV1 = Type.Union([
  Type.Literal("forge_station"),
  Type.Literal("alchemy_furnace"),
  Type.Literal("rogue_village"),
  Type.Literal("mutant_nest"),
  Type.Literal("scroll_hidden"),
  Type.Literal("spirit_herb_valley"),
  Type.Literal("herb_patch"),
  Type.Literal("qi_spring"),
  Type.Literal("trade_spot"),
  Type.Literal("shelter_spot"),
  Type.Literal("water_source"),
]);
export type PoiNoviceKindV1 = Static<typeof PoiNoviceKindV1>;

export const PoiSpawnedEventV1 = Type.Object(
  {
    v: Type.Literal(1),
    kind: Type.Literal("poi_spawned"),
    poi_id: Type.String({ minLength: 1 }),
    poi_type: PoiNoviceKindV1,
    zone: Type.String({ minLength: 1 }),
    pos: Type.Array(Type.Number(), { minItems: 3, maxItems: 3 }),
    selection_strategy: Type.String({ minLength: 1 }),
    qi_affinity: Type.Number({ minimum: -1, maximum: 1 }),
    danger_bias: Type.Integer({ minimum: 0 }),
  },
  { additionalProperties: false },
);
export type PoiSpawnedEventV1 = Static<typeof PoiSpawnedEventV1>;

export const TrespassEventV1 = Type.Object(
  {
    v: Type.Literal(1),
    kind: Type.Literal("trespass"),
    village_id: Type.String({ minLength: 1 }),
    player_id: Type.String({ minLength: 1 }),
    killed_npc_count: Type.Integer({ minimum: 1 }),
    refusal_until_wall_clock_secs: Type.Number({ minimum: 0 }),
  },
  { additionalProperties: false },
);
export type TrespassEventV1 = Static<typeof TrespassEventV1>;

export function validatePoiSpawnedEventV1Contract(data: unknown): ValidationResult {
  return validate(PoiSpawnedEventV1, data);
}

export function validateTrespassEventV1Contract(data: unknown): ValidationResult {
  return validate(TrespassEventV1, data);
}

File: server/src/schema/poi_novice.rs
//! plan-poi-novice-v1 — server 侧 POI novice IPC wire serde 结构。

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum PoiNoviceKindV1 {
    ForgeStation,
    AlchemyFurnace,
    RogueVillage,
    MutantNest,
    ScrollHidden,
    SpiritHerbValley,
    HerbPatch,
    QiSpring,
    TradeSpot,
    ShelterSpot,
    WaterSource,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct PoiSpawnedEventV1 {
    pub v: u8,
    pub kind: String,
    pub poi_id: String,
    pub poi_type: PoiNoviceKindV1,
    pub zone: String,
    pub pos: [f64; 3],
    pub selection_strategy: String,
    pub qi_affinity: f32,
    pub danger_bias: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct TrespassEventV1 {
    pub v: u8,
    pub kind: String,
    pub village_id: String,
    pub player_id: String,
    pub killed_npc_count: u32,
    pub refusal_until_wall_clock_secs: u64,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn poi_spawned_sample_roundtrips() {
        let json =
            include_str!("../../../agent/packages/schema/samples/poi-spawned-event.sample.json");
        let event: PoiSpawnedEventV1 =
            serde_json::from_str(json).expect("poi spawned sample should deserialize");
        assert_eq!(event.v, 1);
        assert_eq!(event.kind, "poi_spawned");
        assert_eq!(event.poi_type, PoiNoviceKindV1::ForgeStation);
        let back = serde_json::to_string(&event).expect("serialize");
        let parsed: PoiSpawnedEventV1 = serde_json::from_str(&back).expect("deserialize");
        assert_eq!(parsed, event);
    }

    #[test]
    fn trespass_sample_roundtrips() {
        let json =
            include_str!("../../../agent/packages/schema/samples/trespass-event.sample.json");
        let event: TrespassEventV1 =
            serde_json::from_str(json).expect("trespass sample should deserialize");
        assert_eq!(event.v, 1);
        assert_eq!(event.kind, "trespass");
        assert_eq!(event.village_id, "spawn:rogue_village");
        let back = serde_json::to_string(&event).expect("serialize");
        let parsed: TrespassEventV1 = serde_json::from_str(&back).expect("deserialize");
        assert_eq!(parsed, event);
    }
}
